classes/auto_grader/average_grade_no_straddle.php

```
@@ -0,0 +1,163 @@
+     *
+     */
+    public function create_auto_grade_if_rules_match() {
+        if (!$this->grades_are_close_enough()) {
+            return;
+        }
+
+        $gradeclassesadminsetting = self::get_config_setting('autogradeclassboundaries');
+        if (!$gradeclassesadminsetting) {
+            return;
+        }
+
+        if ($this->grades_straddle_class_boundaries($gradeclassesadminsetting)) {
+            return;
+        }
+
+        // We also want to do some other checks, e.g. that advanced grading is not being used, before applying auto grade.
+        // Parent method will do all that.
+        parent::create_auto_grade_if_rules_match();
+    }
+
+    /**
+     * Are grades close enough (within % setting in cm settings) to allow auto agreement?
+     * @return bool
+     */
+    private function grades_are_close_enough(): bool {
+        $grades = $this->grades_as_percentages();
+        if (empty($grades)) {
+            return false;
+        }
+        $maxgrade = max($grades);
+        $mingrade = min($grades);
+        return ($maxgrade - $mingrade) <= $this->percentage;
+    }
+
+    /**
+     * Do the grades awarded for this assessment straddle class boundaries?
+     * @param array $gradeclassesadminsetting the grade classes and boundaries as defined by site admin setting.
+     * @return bool
+     */
+    public function grades_straddle_class_boundaries(array $gradeclassesadminsetting): bool {
+        $grades = $this->grades_as_percentages();
+        if (empty($grades)) {
+            // No grades found, so we are not applying this rule.
+            return false;
+        }
+
+        $gradeclassesseen = [];
+        foreach ($grades as $grade) {
+            foreach ($gradeclassesadminsetting as $index => $gradeclassboundaries) {
+                $boundarybottom = $gradeclassboundaries[0];
+                $boundarytop = $gradeclassboundaries[1];
+                if ($grade >= $boundarybottom && $grade <= $boundarytop) {
```

What happens where grades sit between ranges, either due to the ranges being badly defined or the grade having more than two digits after the decimal point - as example, 59.995 would cause a problem here.

---

The validation on data entry will stop this from happening

---

I would say that's too far - the function should throw an exception if it tries to handle a grade that doesn't fit....
But this issue is kind of the core of why I thing we should specify the ranges in the simpler way - right now we're testing to see if a floating point number fits into a series of non-contiguous ranges and have no plan for when it doesn't.

---

Again - maybe this needs a casting vote from someone else.

---

The validation will prevent non-contiguous ranges being entered.  Also note that this setting is hardly ever likely to be changed by the site admin - likely to be set only once, using the example values provided, and then left that way.